Index: configure.in
===================================================================
--- configure.in	(revision 521)
+++ configure.in	(working copy)
@@ -1379,13 +1379,8 @@
 Makefile
 evolution-data-server.pc
 addressbook/Makefile
-addressbook/idl/Makefile
-addressbook/libebook-orbit/Makefile
-addressbook/libebook-orbit/libebook.pc
 addressbook/libebook-dbus/Makefile
 addressbook/libebook-dbus/libebook.pc
-addressbook/libedata-book-orbit/Makefile
-addressbook/libedata-book-orbit/libedata-book.pc
 addressbook/libedata-book-dbus/Makefile
 addressbook/libedata-book-dbus/libedata-book.pc
 addressbook/backends/Makefile
@@ -1396,70 +1391,11 @@
 addressbook/tests/Makefile
 addressbook/tests/ebook/Makefile
 addressbook/tests/vcard/Makefile
-art/Makefile
-calendar/Makefile
-calendar/idl/Makefile
-calendar/libecal/Makefile
-calendar/libecal/libecal.pc
-calendar/libecal-dbus/Makefile
-calendar/libecal-dbus/libecal.pc
-calendar/libedata-cal/Makefile
-calendar/libedata-cal/libedata-cal.pc
-calendar/libedata-cal-dbus/Makefile
-calendar/libedata-cal-dbus/libedata-cal.pc
-calendar/backends/Makefile
-calendar/backends/file/Makefile
-calendar/backends/groupwise/Makefile
-calendar/backends/http/Makefile
-calendar/backends/contacts/Makefile
-calendar/backends/weather/Makefile
-calendar/tests/Makefile
-calendar/tests/ecal/Makefile
 ucamel/Makefile
-camel/Makefile
-camel/providers/Makefile
-camel/providers/groupwise/Makefile
-camel/providers/imap/Makefile
-camel/providers/imapp/Makefile
-camel/providers/imap4/Makefile
-camel/providers/local/Makefile
-camel/providers/nntp/Makefile
-camel/providers/pop3/Makefile
-camel/providers/sendmail/Makefile
-camel/providers/smtp/Makefile
-camel/tests/Makefile
-camel/tests/folder/Makefile
-camel/tests/lib/Makefile
-camel/tests/message/Makefile
-camel/tests/mime-filter/Makefile
-camel/tests/misc/Makefile
-camel/tests/smime/Makefile
-camel/tests/stream/Makefile
-camel/camel.pc
-camel/camel-provider.pc
 libdb/Makefile
 libedataserver/Makefile
 libedataserver/libedataserver.pc
-libedataserverui/Makefile
-libedataserverui/libedataserverui.pc
-servers/Makefile
-servers/groupwise/Makefile
-servers/groupwise/libegroupwise.pc
-servers/exchange/Makefile
-servers/exchange/xntlm/Makefile
-servers/exchange/lib/Makefile
-servers/exchange/storage/Makefile
-servers/exchange/storage/libexchange-storage.pc 
 src/Makefile
-docs/Makefile
-docs/reference/Makefile
-docs/reference/addressbook/Makefile
-docs/reference/addressbook/libebook/Makefile
-docs/reference/calendar/Makefile
-docs/reference/calendar/libecal/Makefile
-docs/reference/calendar/libedata-cal/Makefile
-docs/reference/libedataserver/Makefile
-po/Makefile.in
 ])
 
 case $host in
Index: addressbook/Makefile.am
===================================================================
--- addressbook/Makefile.am	(revision 521)
+++ addressbook/Makefile.am	(working copy)
@@ -1,7 +1,3 @@
-if ENABLE_DBUS
 DIRS=libebook-dbus libedata-book-dbus backends
-else
-DIRS=idl libedata-book-orbit libebook-orbit tests backends
-endif
 
 SUBDIRS = $(DIRS)
Index: Makefile.am
===================================================================
--- Makefile.am	(revision 521)
+++ Makefile.am	(working copy)
@@ -5,24 +5,10 @@
 LIBDB = libdb
 endif
 
-if ENABLE_CALENDAR
-CALENDAR_DIR = calendar
-endif
+SUBDIRS = $(LIBDB) libedataserver ucamel addressbook 
 
-if ENABLE_CAMEL
-CAMEL_DIR = camel
-else
-CAMEL_DIR = ucamel
-endif
+DIST_SUBDIRS= libedataserver ucamel addressbook 
 
-if ENABLE_DBUS
-SUBDIRS = $(LIBDB) libedataserver $(CAMEL_DIR) addressbook $(CALENDAR_DIR) libedataserverui docs art po
-else
-SUBDIRS = $(LIBDB) libedataserver servers $(CAMEL_DIR) addressbook $(CALENDAR_DIR) libedataserverui src docs art po
-endif
-
-DIST_SUBDIRS= libedataserver servers camel ucamel addressbook calendar libedataserverui src docs art po
-
 changelogs =			\
 	ChangeLog
 
