Index: addressbook/libebook-dbus/e-book.c
===================================================================
--- addressbook/libebook-dbus/e-book.c.orig	2008-02-19 11:04:22.000000000 +0000
+++ addressbook/libebook-dbus/e-book.c	2008-02-19 11:07:02.000000000 +0000
@@ -440,7 +440,8 @@
 		char *uri;
 
 		filename = g_build_filename (g_get_home_dir(),
-					     ".evolution/addressbook/local/system",
+					     ".osso-abook",
+					     "db",
 					     NULL);
 		uri = g_strdup_printf ("file://%s", filename);
 
