Index: addressbook/backends/file/e-book-backend-file.c
===================================================================
--- addressbook/backends/file/e-book-backend-file.c.orig	2008-02-19 11:07:09.000000000 +0000
+++ addressbook/backends/file/e-book-backend-file.c	2008-02-19 11:07:23.000000000 +0000
@@ -1200,6 +1200,8 @@
 
 		env->set_errcall (env, file_errcall);
 
+		env->set_cachesize (env, 0, 128 * 1024, 1);
+
 		/* Set the allocation routines to the non-aborting GLib functions */
 		env->set_alloc (env, (void *(*)(size_t))g_try_malloc, 
 				(void *(*)(void *, size_t))g_try_realloc,
