Disable i18n in the builds.

Index: git/configure.in
===================================================================
--- git.orig/configure.in	2008-04-24 14:15:59.000000000 +0100
+++ git/configure.in	2008-04-24 14:19:34.000000000 +0100
@@ -120,10 +120,10 @@
 AC_SUBST(HAVE_JW)
 
 dnl I18N stuff
-AC_PROG_INTLTOOL
+dnl AC_PROG_INTLTOOL
 
-ALL_LINGUAS="am ar az be bg bn bs ca cs cy da de dz el en_AU en_CA en_GB es et eu fa fi fr ga gl gu he hi hr hu id is it ja ko lt lv mk ml mn ms nb ne nl nn no pa pl pt pt_BR ro ru rw sk sl sq sr sr@Latn sv tr uk vi wa xh zh_CN zh_TW"
-AM_GLIB_GNU_GETTEXT
+dnl ALL_LINGUAS="am ar az be bg bn bs ca cs cy da de dz el en_AU en_CA en_GB es et eu fa fi fr ga gl gu he hi hr hu id is it ja ko lt lv mk ml mn ms nb ne nl nn no pa pl pt pt_BR ro ru rw sk sl sq sr sr@Latn sv tr uk vi wa xh zh_CN zh_TW"
+dnl AM_GLIB_GNU_GETTEXT
 
 GETTEXT_PACKAGE=evolution-data-server-$BASE_VERSION
 AC_SUBST(GETTEXT_PACKAGE)
@@ -1450,7 +1450,6 @@
 docs/reference/calendar/libecal/Makefile
 docs/reference/calendar/libedata-cal/Makefile
 docs/reference/libedataserver/Makefile
-po/Makefile.in
 ])
 
 case $host in
Index: git/Makefile.am
===================================================================
--- git.orig/Makefile.am	2008-04-24 14:19:39.000000000 +0100
+++ git/Makefile.am	2008-04-24 14:19:50.000000000 +0100
@@ -36,9 +36,6 @@
 	NEWS						\
 	marshal.mk					\
 	iconv-detect.c					\
-	intltool-merge.in				\
-	intltool-update.in				\
-	intltool-extract.in				\
 	$(pkgconfig_DATA:-$(API_VERSION).pc=.pc.in) 	\
 	autogen.sh					\
 	$(LIBDB_FILES)
