Remove subdirectories we don't use.

Index: git/Makefile.am
===================================================================
--- git.orig/Makefile.am	2008-04-24 14:29:36.000000000 +0100
+++ git/Makefile.am	2008-04-24 14:34:14.000000000 +0100
@@ -10,7 +10,7 @@
 endif
 
 if ENABLE_DBUS
-SUBDIRS = $(LIBDB) libedataserver camel addressbook $(CALENDAR_DIR) libedataserverui docs art po
+SUBDIRS = $(LIBDB) libedataserver addressbook docs
 else
 SUBDIRS = $(LIBDB) libedataserver servers camel addressbook $(CALENDAR_DIR) libedataserverui src docs art po
 endif
