Index: git-foo/addressbook/libedata-book-dbus/e-data-book-view.c
===================================================================
--- git-foo.orig/addressbook/libedata-book-dbus/e-data-book-view.c	2008-05-16 11:29:03.000000000 +0100
+++ git-foo/addressbook/libedata-book-dbus/e-data-book-view.c	2008-05-16 12:26:04.000000000 +0100
@@ -146,10 +146,17 @@
 book_destroyed_cb (gpointer data, GObject *dead)
 {
   EDataBookView *view = E_DATA_BOOK_VIEW (data);
+  EDataBookViewPrivate *priv = view->priv;
+
   /* The book has just died, so unset the pointer so we don't try and remove a
      dead weak reference. */
   view->priv->book = NULL;
-  g_object_unref (view);
+
+  /* If the view is running stop it here. */
+  if (priv->running) {
+    e_book_backend_stop_book_view (priv->backend, view);
+    priv->running = FALSE;
+  }
 }
 
 /* Oh dear god, think of the children. */
@@ -233,11 +240,6 @@
     priv->book = NULL;
   }
 
-  if (priv->running) {
-    e_book_backend_stop_book_view (priv->backend, book_view);
-    priv->running = FALSE;
-  }
-
   if (priv->backend) {
     e_book_backend_remove_book_view (priv->backend, book_view);
     g_object_unref (priv->backend);
