Index: evolution-data-server.build/addressbook/libebook-dbus/e-book-query.c
===================================================================
--- evolution-data-server.build.orig/addressbook/libebook-dbus/e-book-query.c	2008-08-11 12:12:27.000000000 +0100
+++ evolution-data-server.build/addressbook/libebook-dbus/e-book-query.c	2008-08-11 12:12:32.000000000 +0100
@@ -424,11 +424,14 @@
 {
 	GList **list = data;
 	ESExpResult *r;
+	EBookQuery *term;
 
 	/* just replace the head of the list with the NOT of it. */
-	if (argc > 0) {
-		EBookQuery *term = (*list)->data;
-		(*list)->data = e_book_query_not (term, TRUE);
+	if (argc == 1) {
+		if (*list) {
+			term = (*list)->data;
+			(*list)->data = e_book_query_not (term, TRUE);
+		}
 	}
 
 	r = e_sexp_result_new(f, ESEXP_RES_BOOL);
