Index: evolution-data-server/addressbook/libebook-dbus/e-contact.c
===================================================================
--- evolution-data-server.orig/addressbook/libebook-dbus/e-contact.c	2008-08-28 19:55:41.000000000 +0300
+++ evolution-data-server/addressbook/libebook-dbus/e-contact.c	2008-08-28 19:55:41.000000000 +0300
@@ -270,13 +270,14 @@
 	ATTR_TYPE_STR_FIELD (E_CONTACT_IM_GADUGADU_WORK_3,  EVC_X_GADUGADU,  "im-gadugadu-work-3",  N_("Gadu-Gadu Work Id 3"), FALSE, "WORK", 2),
 	MULTI_LIST_FIELD (E_CONTACT_IM_GADUGADU,  EVC_X_GADUGADU,  "im-gadugadu", N_("Gadu-Gadu Id List"), FALSE),
 
-	/* For TEL */
- 	STRING_FIELD (E_CONTACT_PHONE_TELEPHONE,        EVC_TEL,       "tel",         N_("Telephone"),  FALSE),
+	/* Geo information */
+	STRUCT_FIELD	(E_CONTACT_GEO,  EVC_GEO, "geo",  N_("Geographic Information"),  FALSE, geo_getter, geo_setter, e_contact_geo_get_type),
 
-	MULTI_LIST_FIELD (E_CONTACT_SIP, 	  EVC_X_SIP,    "sip",    N_("SIP address"),          FALSE),
+	/* For TEL */
+	MULTI_LIST_FIELD     (E_CONTACT_TEL,      EVC_TEL,        "phone",      N_("Telephone"),      FALSE),
 
-	/* Geo information */
-	STRUCT_FIELD	(E_CONTACT_GEO,  EVC_GEO, "geo",  N_("Geographic Information"),  FALSE, geo_getter, geo_setter, e_contact_geo_get_type)
+	/* For SIP */
+	MULTI_LIST_FIELD (E_CONTACT_SIP, 	  EVC_X_SIP,    "sip",    N_("SIP address"),          FALSE)
 };
 
 #undef LIST_ELEM_STR_FIELD
Index: evolution-data-server/addressbook/libebook-dbus/e-contact.h
===================================================================
--- evolution-data-server.orig/addressbook/libebook-dbus/e-contact.h	2008-08-28 19:55:41.000000000 +0300
+++ evolution-data-server/addressbook/libebook-dbus/e-contact.h	2008-08-28 19:55:41.000000000 +0300
@@ -195,11 +195,11 @@
 
 	E_CONTACT_IM_GADUGADU,   /* Multi-valued */
 	
-	E_CONTACT_PHONE_TELEPHONE,
+	E_CONTACT_GEO,	/* structured field (EContactGeo) */
 
-	E_CONTACT_SIP,
+	E_CONTACT_TEL,
 
-	E_CONTACT_GEO,	/* structured field (EContactGeo) */	
+	E_CONTACT_SIP,
 
 	E_CONTACT_FIELD_LAST,
 	E_CONTACT_FIELD_FIRST        = E_CONTACT_UID,
Index: evolution-data-server/addressbook/libedata-book-dbus/e-book-backend-sexp.c
===================================================================
--- evolution-data-server.orig/addressbook/libedata-book-dbus/e-book-backend-sexp.c	2008-08-06 13:48:36.000000000 +0300
+++ evolution-data-server/addressbook/libedata-book-dbus/e-book-backend-sexp.c	2008-08-28 21:08:59.000000000 +0300
@@ -131,17 +131,16 @@
 compare_phone (EContact *contact, const char *str,
 	       char *(*compare)(const char*, const char*))
 {
-	int i;
+	GList *phones;
 	gboolean rv = FALSE;
 
-	for (i = E_CONTACT_FIRST_PHONE_ID; i <= E_CONTACT_LAST_PHONE_ID; i ++) {
-		char *phone = e_contact_get (contact, i);
-
-		rv = phone && compare(phone, str);
-		g_free (phone);
-
-		if (rv)
-			break;
+	phones = e_contact_get (contact, E_CONTACT_TEL);
+	while (phones) {
+		char *phone = phones->data;
+
+		phones = g_list_delete_link(phones, phones);
+		rv = rv || (phone && compare(phone, str));
+		g_free(phone);
 	}
 
 	return rv;
