diff -urNpb evolution-data-server.org/addressbook/libebook-dbus/e-contact.c evolution-data-server/addressbook/libebook-dbus/e-contact.c
--- evolution-data-server.org/addressbook/libebook-dbus/e-contact.c	2008-09-15 16:16:51.000000000 +0300
+++ evolution-data-server/addressbook/libebook-dbus/e-contact.c	2008-09-15 16:21:10.000000000 +0300
@@ -277,7 +277,15 @@ static const EContactFieldInfo field_inf
 	MULTI_LIST_FIELD     (E_CONTACT_TEL,      EVC_TEL,        "phone",      N_("Telephone"),      FALSE),
 
 	/* For SIP */
-	MULTI_LIST_FIELD (E_CONTACT_SIP, 	  EVC_X_SIP,    "sip",    N_("SIP address"),          FALSE)
+	MULTI_LIST_FIELD (E_CONTACT_SIP, 	  EVC_X_SIP,    "sip",    N_("SIP address"),          FALSE),
+
+	ATTR_TYPE_STR_FIELD (E_CONTACT_IM_SKYPE_HOME_1,    EVC_X_SKYPE,    "im-skype-home-1",    N_("Skype Home Id 1"),             FALSE, "HOME", 0),
+	ATTR_TYPE_STR_FIELD (E_CONTACT_IM_SKYPE_HOME_2,    EVC_X_SKYPE,    "im-skype-home-2",    N_("Skype Home Id 2"),             FALSE, "HOME", 1),
+	ATTR_TYPE_STR_FIELD (E_CONTACT_IM_SKYPE_HOME_3,    EVC_X_SKYPE,    "im-skype-home-3",    N_("Skype Home Id 3"),             FALSE, "HOME", 2),
+	ATTR_TYPE_STR_FIELD (E_CONTACT_IM_SKYPE_WORK_1,    EVC_X_SKYPE,    "im-skype-work-1",    N_("Skype Work Id 1"),             FALSE, "WORK", 0),
+	ATTR_TYPE_STR_FIELD (E_CONTACT_IM_SKYPE_WORK_2,    EVC_X_SKYPE,    "im-skype-work-2",    N_("Skype Work Id 2"),             FALSE, "WORK", 1),
+	ATTR_TYPE_STR_FIELD (E_CONTACT_IM_SKYPE_WORK_3,    EVC_X_SKYPE,    "im-skype-work-3",    N_("Skype Work Id 3"),             FALSE, "WORK", 2),
+	MULTI_LIST_FIELD (E_CONTACT_IM_SKYPE,  	  EVC_X_SKYPE,     "im-skype",     N_("Skype Id List"),             FALSE)
 };
 
 #undef LIST_ELEM_STR_FIELD
diff -urNpb evolution-data-server.org/addressbook/libebook-dbus/e-contact.h evolution-data-server/addressbook/libebook-dbus/e-contact.h
--- evolution-data-server.org/addressbook/libebook-dbus/e-contact.h	2008-09-15 16:16:52.000000000 +0300
+++ evolution-data-server/addressbook/libebook-dbus/e-contact.h	2008-09-15 16:21:10.000000000 +0300
@@ -201,6 +201,14 @@ typedef enum {
 
 	E_CONTACT_SIP,
 
+	E_CONTACT_IM_SKYPE_HOME_1,     /* Synthetic string field */
+	E_CONTACT_IM_SKYPE_HOME_2,     /* Synthetic string field */
+	E_CONTACT_IM_SKYPE_HOME_3,     /* Synthetic string field */
+	E_CONTACT_IM_SKYPE_WORK_1,     /* Synthetic string field */
+	E_CONTACT_IM_SKYPE_WORK_2,     /* Synthetic string field */
+	E_CONTACT_IM_SKYPE_WORK_3,     /* Synthetic string field */
+	E_CONTACT_IM_SKYPE,    	       /* Multi-valued */
+
 	E_CONTACT_FIELD_LAST,
 	E_CONTACT_FIELD_FIRST        = E_CONTACT_UID,
 
diff -urNpb evolution-data-server.org/addressbook/libebook-dbus/e-vcard.h evolution-data-server/addressbook/libebook-dbus/e-vcard.h
--- evolution-data-server.org/addressbook/libebook-dbus/e-vcard.h	2008-09-15 16:16:52.000000000 +0300
+++ evolution-data-server/addressbook/libebook-dbus/e-vcard.h	2008-09-15 16:20:25.000000000 +0300
@@ -90,6 +90,7 @@ G_BEGIN_DECLS
 #define EVC_X_GADUGADU        "X-GADUGADU"
 #define EVC_X_GROUPWISE     	"X-GROUPWISE"
 #define EVC_X_BOOK_URI     	"X-EVOLUTION-BOOK-URI"
+#define EVC_X_SKYPE           "X-SKYPE"
 
 typedef enum {
 	EVC_FORMAT_VCARD_21,
diff -urNpb evolution-data-server.org/addressbook/libedata-book-dbus/e-book-backend-sexp.c evolution-data-server/addressbook/libedata-book-dbus/e-book-backend-sexp.c
--- evolution-data-server.org/addressbook/libedata-book-dbus/e-book-backend-sexp.c	2008-09-15 16:16:52.000000000 +0300
+++ evolution-data-server/addressbook/libedata-book-dbus/e-book-backend-sexp.c	2008-09-15 16:21:01.000000000 +0300
@@ -112,6 +112,13 @@ compare_im_groupwise (EContact *contact,
 }
 
 static gboolean
+compare_im_skype (EContact *contact, const char *str,
+		      char *(*compare)(const char*, const char*))
+{
+	return compare_im (contact, str, compare, E_CONTACT_IM_SKYPE);
+}
+
+static gboolean
 compare_email (EContact *contact, const char *str,
 	       char *(*compare)(const char*, const char*))
 {
@@ -269,6 +276,7 @@ static struct prop_info {
 	LIST_PROP ( "im-jabber", compare_im_jabber ),
 	LIST_PROP ( "im-yahoo",  compare_im_yahoo ),
 	LIST_PROP ( "im-groupwise", compare_im_groupwise ),
+	LIST_PROP ( "im-skype", compare_im_skype ),
 	LIST_PROP ( "email",     compare_email ),
 	LIST_PROP ( "phone",     compare_phone ),
 	LIST_PROP ( "address",   compare_address ),
